skycfg.go

```diff
@@ -149,25 +149,48 @@ func WithProtoRegistry(r unstableProtoRegistry) LoadOption {
 	})
 }
 
+// UnstablePredeclaredModules returns a Starlark string dictionary with
+// predeclared Skycfg modules which can be used in starlark.ExecFile.
+//
+// Takes in unstableProtoRegistry as param (if nil will use standard proto
+// registry).
+//
+// Currently provides these modules (see REAMDE for more detailed description):
+//  * fail   - interrupts execution and prints a stacktrace.
+//  * hash   - supports md5, sha1 and sha245 functions.
+//  * json   - marshals plain values (dicts, lists, etc) to JSON.
+//  * proto  - package for constructing Protobuf messages.
+//  * struct - experimental Starlark struct support.
+//  * yaml   - same as "json" package but for YAML.
+//  * url    - utility package for encoding URL query string.
+func UnstablePredeclaredModules(r unstableProtoRegistry) starlark.StringDict {
+	return starlark.StringDict{
+		"fail":   starlark.NewBuiltin("fail", skyFail),
+		"hash":   impl.HashModule(),
+		"json":   impl.JsonModule(),
+		"proto":  impl.NewProtoModule(r),
+		"struct": starlark.NewBuiltin("struct", starlarkstruct.Make),
+		"yaml":   impl.YamlModule(),
+		"url":    impl.UrlModule(),
+	}
+}
+
 // Load reads a Skycfg config file from the filesystem.
 func Load(ctx context.Context, filename string, opts ...LoadOption) (*Config, error) {
-	protoModule := impl.NewProtoModule(nil /* TODO: registry from options */)
 	parsedOpts := &loadOptions{
-		globals: starlark.StringDict{
-			"fail":   starlark.NewBuiltin("fail", skyFail),
-			"hash":   impl.HashModule(),
-			"json":   impl.JsonModule(),
-			"proto":  protoModule,
-			"struct": starlark.NewBuiltin("struct", starlarkstruct.Make),
-			"yaml":   impl.YamlModule(),
-			"url":    impl.UrlModule(),
-		},
+		globals:    UnstablePredeclaredModules(nil),
 		fileReader: LocalFileReader(filepath.Dir(filename)),
 	}
+
 	for _, opt := range opts {
 		opt.applyLoad(parsedOpts)
 	}
-	protoModule.Registry = parsedOpts.protoRegistry
+
+	// Set the proto registry option if proto package was configured.
+	if p, ok := parsedOpts.globals["proto"]; ok {
+		p.(*impl.ProtoModule).Registry = parsedOpts.protoRegistry
```

Comment:
This will crash if the user set `proto` to something else.

Reply:
Hm, that's right - one could override `proto` key...

```
 WithGlobals adds *additional* global symbols
```

^ should we allow overrides though?

Reply:
We currently allow overrides to provide users an escape hatch, and for stubbing in tests. Even if we decide to forbid it, the user-visible behavior should be an error message instead of a type-conversion panic.

Reply:
> Even if we decide to forbid it, the user-visible behavior should be an error message instead of a type-conversion panic.

Yeah, totally. Was just thinking what's the best way to tackle this.

Reply:
If constructing the default globals is moved into a helper, then we can safely provide both the new interface (returning a single dict) and support the existing `Load()` semantics.